### PR TITLE
Reverse string

### LINKED ISSUE #23: Reverse_String Challenge
"""
## Description

Write a function that reverses a given string.

---

## Behavior

1. **Take a string** as input.
2. **Return the reversed string** as output.
3. Handle edge cases such as empty strings and single-character strings.

---

### Example 1
Input: "hello"  
Output: "olleh"

### Example 2
Input: "Python"  
Output: "nohtyP"

### PR DESCRIPTION
**Pull request by marshar007007**
---
Reverse String 
Python Challenge
---

## Behavior

Take a string as input.
Return the reversed string as output.
Handle edge cases such as empty strings and single-character strings.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/reverse_string [hunk 1]
```diff
@@ -0,0 +1,31 @@
+"""
```
Comment (hamid4231) [inline solutions/reverse_string:4]: Add these two lines at the beginning of your main Python file and test file.
#!/usr/bin/env python3
#-*- coding: utf-8 -*-

File: solutions/tests/test_reverse_string [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+Unit tests for the reverse_string function.
+"""
+import unittest
+from reverse_string import reverse_string
+
+class TestReverseString(unittest.TestCase):
+    """Tests for the reverse_string function."""
+
+    def test_reverse_normal_string(self):
+        """Should reverse a normal string."""
```
Comment (hamid4231) [inline solutions/tests/test_reverse_string]: @marshar007007 you can rephrase the docstring for each test and remove the should. In this case for example: "Reverses a normal string"

File: solutions/tests/test_reverse_string [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    def test_reverse_empty_string(self):
+        """Should return an empty string when input is empty."""
+        self.assertEqual(reverse_string(""), "")
+
+    def test_reverse_single_character(self):
+        """Should return the same character for single character input."""
+        self.assertEqual(reverse_string("a"), "a")
+
+    def test_reverse_raises_type_error(self):
```
Comment (hamid4231) [inline solutions/tests/test_reverse_string:44]: @marshar007007 I think it would be helpful to test one or two more normal strings and then separate the normal cases from defensive and boundary cases using comments. you should also add one or two more defensive test cases.

File: solutions/reverse_string [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    Examples:
+        >>> reverse_string("hello")
+        'olleh'
+        >>> reverse_string("")
+        ''
+        >>> reverse_string("a")
+        'a'
+    """
+    if not isinstance(input_string, str):
```
Comment (hamid4231) [inline solutions/reverse_string:33]: @marshar007007 I think it would be helpful to use comments here to describe the strategy before your implementation.
For example: "# Validates the input and if the input is valid reverses the input"

**Review by hamid4231 — CHANGES_REQUESTED** [review]

File: solutions/reverse_string (hunk 1 above)
Comment (marshar007007) [reply solutions/reverse_string:4]: thank you Hamid! I made added to both files based on your suggestion.

File: solutions/reverse_string (hunk 4 above)
Comment (marshar007007) [reply solutions/reverse_string:33]: Done correction!

File: solutions/tests/test_reverse_string (hunk 2 above)
Comment (marshar007007) [reply solutions/tests/test_reverse_string]: Alright! I will remove extra "should"

File: solutions/tests/test_reverse_string (hunk 3 above)
Comment (marshar007007) [reply solutions/tests/test_reverse_string:44]: I divide cases into normal and defensive cases using comments separately. I also added more test cases to check. I think I would be fine now.  Thank you for your feedbacks.

File: solutions/reverse_string (hunk 4 above)
Comment (hamid4231) [reply solutions/reverse_string:33]: please remove the "and" and use comma instead for better punctuation.
"# Validates the input and if the input is valid, reverses the input"

**Review by hamid4231 — CHANGES_REQUESTED** [review]
Hello @marshar007007, the rest looks fine. I have suggested a little change in your implementation comment. And also, It would be great if you can separate defensive and boundary cases.

File: solutions/reverse_string (hunk 4 above)
Comment (marshar007007) [reply solutions/reverse_string:33]: thank u, I corrected it !

**Review by hamid4231 — APPROVED** [review]
Hello @marshar007007, your solution looks great and it is ready to be merged. I will merge your solution to the main branch and delete your challenge's branch. Thank you for your collaboration and patience.
